[ADD] l10n_ar: Liquido Product document management

zaoral opened the pull request:
* Add refunds with liquido docs
* Add Is AFIP POS field (l10n_ar_is_pos)
* Adapt letter mapping

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

vbe-odoo commented on addons/l10n_ar/models/account_move.py:
```diff
@@ -34,6 +34,17 @@ def _l10n_ar_get_document_number_parts(self, document_number, document_type_code
+        # in that case, the user can create a new Sales Liquido Producto Journal
+
+        # When issuer is supplie the numbering works opposite (supplier enumerate the invoices, and the customer encode the bill)
+        if self.country_code == 'AR' and self.l10n_latam_use_documents and ((self.journal_id.type == 'sale' and not self.journal_id.l10n_ar_is_pos) or (
```
@zaoral update the condition here if POS always electronic

zaoral replied:
Hi @vbe-odoo 
I have applied the changes but not sure if they are the right changes. I understood/applied the changes taking into account that is it a Pos System, do not take into account if it is electronic or not. Please let me know if you think we should take into account if it is electronic.
Wait your feedback

Best Regards

vbe-odoo replied:
Hi @zaoral !
Dismiss my message about the electronic hehe! If l10n_ar_is_pos = True (selected) then it should work as a regular invoice (taking the number from AFIP and not being manual)
Cheers!

vbe-odoo replied:
So it should work now, as you took into account the POS.
I can test! 🚀

vbe-odoo replied:
Hello @zaoral can you help me do a rebase on this PR?
As there is an issue that my be solved with a rebase or if not if you can help me check what's happening.
So, when installing AR there is a conflict with some demos data DOCs and then it does not install the demo data, see the screenshot below:
<img width="995" alt="image" src="https://user-images.githubusercontent.com/45669815/215124990-c35ef204-8369-4dc1-8727-bf189e2b05ee.png">

Thanks!

vin-odoo commented on addons/l10n_ar/models/account_journal.py:
```diff
@@ -112,16 +136,22 @@ def _check_afip_configurations(self):
+    @api.constrains('l10n_ar_afip_pos_system')
+    def _check_afip_pos_system(self):
+        journals = self.filtered(
+            lambda x: x.l10n_ar_is_pos and x.type == 'purchase' and
+            self.l10n_ar_afip_pos_system not in ['II_IM', 'RLI_RLM', 'RAW_MAW'])
```
```suggestion
            lambda j: j.l10n_ar_is_pos and j.type == 'purchase' and
            j.l10n_ar_afip_pos_system not in ['II_IM', 'RLI_RLM', 'RAW_MAW'])
```

vin-odoo commented on addons/l10n_ar/models/account_journal.py:
```diff
@@ -112,16 +136,22 @@ def _check_afip_configurations(self):
-            x.company_id.account_fiscal_country_id.code == "AR")
-
-        if to_review.filtered(lambda x: x.l10n_ar_afip_pos_number == 0):
+        if self.filtered(lambda x: x.l10n_ar_is_pos and x.l10n_ar_afip_pos_number == 0):
```
```suggestion
        if self.filtered(lambda j: j.l10n_ar_is_pos and j.l10n_ar_afip_pos_number == 0):
```
I don't mind using a single letter in the filter, but I prefer it to be the first of what we are iterating on. (Journals here)

vin-odoo commented on addons/l10n_ar/models/account_journal.py:
```diff
@@ -112,16 +136,22 @@ def _check_afip_configurations(self):
             raise ValidationError(_('Please define an AFIP POS number'))
 
-        if to_review.filtered(lambda x: x.l10n_ar_afip_pos_number > 99999):
+        if self.filtered(lambda x: x.l10n_ar_is_pos and x.l10n_ar_afip_pos_number > 99999):
```
```suggestion
        if self.filtered(lambda j: j.l10n_ar_is_pos and j.l10n_ar_afip_pos_number > 99999):
```

zaoral replied:
this one was also tested and applied

vin-odoo commented on addons/l10n_ar/models/account_move.py:
```diff
@@ -99,10 +109,8 @@ def _get_l10n_latam_documents_domain(self):
-            if codes:
-                domain.append(('code', 'in', codes))
-            if self.move_type == 'in_refund':
+            domain += self.journal_id._get_journal_codes_domain()
```
Would be better to replace this part and user expressions, such as `expression.AND`

vbe-odoo replied:
@zaoral ping

zaoral replied:
we change the approach to use expression.AND

vin-odoo reviewed (changes requested):
Did a first tech review. I will probably let @jco-odoo check too as he is a lot more familiar with l10n_ar.

zaoral commented:
Hi @vin-odoo 
Thank you for the review.
I have directly applied some or just suggestions and other ones will stay open mind while I review then more close it

Best Regards

jco-odoo commented on addons/l10n_ar/models/account_move.py:
```diff
@@ -34,6 +34,25 @@ def _l10n_ar_get_document_number_parts(self, document_number, document_type_code
+
+        # NOTE: There is a corner case where 2 sales documents can have the same number for the same DOC from a
+        # different vendor, in that case, the user can create a new Sales Liquido Producto Journal
+        return True if self.l10n_latam_use_documents and (
+            (self.journal_id.type == 'purchase' and not self.journal_id.l10n_ar_is_pos) or
+            (self.journal_id.type == 'sale' and not self.journal_id.l10n_ar_is_pos)
+        ) else False
```
```suggestion
        return self.l10n_latam_use_documents and (
            (self.journal_id.type == 'purchase' and not self.journal_id.l10n_ar_is_pos) or
            (self.journal_id.type == 'sale' and not self.journal_id.l10n_ar_is_pos)
        ) 
```

vbe-odoo replied:
@zaoral ping

zaoral replied:
push the last version with the simpler if, and remove the f/else condition

zaoral replied:
add unit tests

zaoral replied:
we simplify the method as we mentioned in the last call. and also added unit tests

vbe-odoo commented:
@zaoral 
I've done the FUN test and now it works OK and looks good!

Also, all runbot tests - expect style. - are green! :🎉 

@jco-odoo this is ready for technical review.
Thanks!
cc' @fvz-odoo @ren-odoo

zaoral commented:
Hi @vbe-odoo 

Thank you for the review.

I have updated the pr to the last changes we agreed on, resolved the pending runbot error for styling, and also added the unit test for liquido producto sales/purchase to ensure when we are asking the manual numbering or not taking into account manual numbering and is AFIP POS new field.

Is ready for your review @jco-odoo
both this one and https://github.com/odoo/enterprise/pull/36322

Best Regards

vbe-odoo commented:
Thanks for the changes and the ping @zaoral 

I've done the FUN tests and LGTM, and this is ready for technical review @jco-odoo 
This PR has another one linked that is: https://github.com/odoo/enterprise/pull/36322
Thanks!
cc' @ren-odoo @fvz-odoo

jco-odoo commented on addons/l10n_ar/models/account_journal.py:
```diff
@@ -9,14 +9,27 @@ class AccountJournal(models.Model):
+    @api.depends('l10n_ar_is_pos')
+    def _compute_l10n_ar_afip_pos_system(self):
+        for journal in self:
+            journal.l10n_ar_afip_pos_system = journal.l10n_ar_is_pos and journal.l10n_ar_afip_pos_system
```
This is a weird compute.  Why not an onchange to set values back?  Or is this the new style?

jco-odoo commented on addons/l10n_ar/models/account_journal.py:
```diff
@@ -65,18 +78,16 @@ def _get_journal_letter(self, counterpart_partner=False):
         return letters
 
-    def _get_journal_codes(self):
+    def _get_journal_codes_domain(self):
```
Why do we have a separate _get_journal_codes_domain and _get_codes_per_journal_type method then?

jco-odoo commented on addons/l10n_ar/models/account_move.py:
```diff
@@ -99,10 +117,11 @@ def _get_l10n_latam_documents_domain(self):
+                self.journal_id._get_journal_codes_domain(),
+            ])
+            if self.move_type in ['out_refund', 'in_refund']:
                 domain = ['|', ('code', 'in', self._get_l10n_ar_codes_used_for_inv_and_ref())] + domain
```
so why not use expression.OR here then?

jco-odoo reviewed (commented):
To check.

jco-odoo commented:
@robodoo r+

jco-odoo commented:
@zaoral Please check for the changes still afterwards, but this way, at least we have it.
